File: bytopic/matrix/java/group2/SetMatrixZeroes.java
// Given an m x n integer matrix matrix, if an element is 0, set its entire row and column to 0's.
// You must do it in place.
// Example:
// Input: matrix = [[1,1,1],[1,0,1],[1,1,1]]
// Output: [[1,0,1],[0,0,0],[1,0,1]]
// Note: Visualize to understand
// Tag: 37/150

import java.util.Arrays;
class SettingMatrixElemToZeroes {
    public static void main(String[] args) {
        int[][] matrix = {
                {1,1,1},
                {1,0,1},
                {1,1,1}
        };

        setZeroes(matrix);
        for(int[] row: matrix)
            System.out.println(Arrays.toString(row));
    }

    static void setZeroes(int[][] matrix) {
        int m = matrix.length, n = matrix[0].length;
        int rowsArray[] = new int[m];
        int colsArray[] = new int[n];

        Arrays.fill(rowsArray, 1);
        Arrays.fill(colsArray, 1);

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (matrix[i][j] == 0) {
                    rowsArray[i] = 0;
                    colsArray[j] = 0;
                }
            }
        }

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (rowsArray[i] == 0 || colsArray[j] == 0)
                    matrix[i][j] = 0;
            }
        }
    }
}

// Method:
// we can use two separate arrays one for rows (rowsArray) and one for columns (colsArray) and initialize them to 1
// while traversing the given matrix whenever we encounter 0 at (i,j), we will set rowsArray[i]=0 and colsArray[j]=0
// After completion of step 2, again iterate through the matrix and for any (i,j), if rowsArray[i] or colsArray[j]
// is 0 then update matrix[i][j] to 0.
// Complexity:
// Time complexity ~ O(mn)
// Space complexity ~ O(m+n)

File: bytopic/matrix/python/group2/set_matrix_zeroes.py
"""
Given an m x n integer matrix matrix, if an element is 0, set its entire row and column to 0's.
You must do it in place.

Example:
Input: matrix = [[1,1,1],[1,0,1],[1,1,1]]
Output: [[1,0,1],[0,0,0],[1,0,1]]
Note: Visualize to understand

Tag: 37/150
Tag: 73/2927, R91/2936 (overall frequency ranking)
"""


def set_zeroes(matrix: list[list[int]]) -> None:
    m = len(matrix)
    n = len(matrix[0])
    first_row_has_zero = False
    first_col_has_zero = False
    # Iterate through matrix to mark the zero row and cols
    for row in range(m):
        for col in range(n):
            if matrix[row][col] == 0:
                if row == 0:
                    first_row_has_zero = True
                if col == 0:
                    first_col_has_zero = True
                matrix[row][0] = matrix[0][col] = 0
    # Iterate through matrix to update the cell to be zero if it's in a zero row or col
    for row in range(1, m):
        for col in range(1, n):
            matrix[row][col] = 0 if matrix[0][col] == 0 or matrix[row][0] == 0 else matrix[row][col]
    # Update the first row and col if they are zero
    if first_row_has_zero:
        for col in range(n):
            matrix[0][col] = 0
    if first_col_has_zero:
        for row in range(m):
            matrix[row][0] = 0


def main():
    matrix = [[1, 1, 1], [1, 0, 1], [1, 1, 1]]
    set_zeroes(matrix)
    print(matrix)


if __name__ == "__main__":
    main()

"""
Time complexity for all three progression is O(m * n).

Space: O(1) for modification in place and using the first row and first col to keep track of zeros instead of 
zeroes_row and zeroes_col
"""
